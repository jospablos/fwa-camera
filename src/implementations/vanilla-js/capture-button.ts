import './capture-button.css';

export interface ButtonProps {
  label: string;
}

function onClick() {
  console.log('click')
}

export const createCaptureButton = ({label}: ButtonProps) => {
  const btn = document.createElement('button');
  btn.type = 'button';
  btn.innerText = label;
  btn.addEventListener('click', onClick);

  btn.className = 'capture-button';

  return btn;
};
