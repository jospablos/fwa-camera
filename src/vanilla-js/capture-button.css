.centered {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.bg--gray {
  background-color: lightgray;
}

.CameraContainer {
  position: relative;
  height: 240px;
  width: 360px;
}

.CameraContainer img {
  width: 100%;
  height: 100%;
}

.CameraContainer video {
  height: 240px;
  width: 360px;
}
