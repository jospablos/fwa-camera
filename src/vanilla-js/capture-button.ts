import { tryScreenshot } from '../core/camera';
import './capture-button.css';

export interface ButtonProps {
  label: string;
  onClick: () => void;
}

const createCaptureFunction = (onError, videoElement) => () => {
  tryScreenshot({
    video: true
  }, { onError, videoElement });
}

function onError(error: string) {
  const messageBox = document.getElementById('capture-button-error');
  if (messageBox) {
    messageBox.innerText = error;
  }
}

const createMessageBox = () => {
  const messageBox = document.createElement('span');
  messageBox.id = 'capture-button-error';
  return messageBox;
}

const createCaptureButton = ({label, onClick}: ButtonProps) => {
  const btn = document.createElement('button');
  
  btn.type = 'button';
  btn.innerText = label;
  btn.addEventListener('click', onClick);

  btn.className = 'capture-button';

  return btn;
};

const createVideoElement = () => {
  const video = document.createElement('video');
  video.style = "display: none;"
  return video;
}
const createCanvasElement = () => {
  const canvas = document.createElement('canvas');
  canvas.style = "display: none;"
  return canvas;
}

const joinElements = (...elements: HTMLElement[]) => {
  const div = document.createElement('div');
  elements.forEach(e => div.appendChild(e));
  return div
}

const paintOnCanvas = (canvas, video) => {
  let ctx = canvas.getContext("2d");

  const width = 320;
  const height = 240;
  canvas.width = width;
  canvas.height = height;

  return setInterval(() => {
    ctx.drawImage(video, 0, 0, width, height);
  }, 200);
}

const createScreenshotHolder = () => {
  const screenshotHolder = document.createElement('img');
  return screenshotHolder;
}

const updateScreenshot = (data, screenshotHolder) => {
  screenshotHolder.src = data;
}

const takeScreenshot = (canvas, screenshotHolder) => () => {
  const data = canvas.toDataURL("image/jpeg");

  updateScreenshot(data, screenshotHolder)

  return canvas.insertAdjacentElement('afterend', screenshotHolder);
};

export const startCapture = (targetElement: Node) => {
  const video = createVideoElement();
  const canvas = createCanvasElement();
  const screenshotHolder = createScreenshotHolder();
  const captureFn = createCaptureFunction(onError, video);
  paintOnCanvas(canvas, video);

  const captureButton = createCaptureButton({label: 'Capture screenshot', onClick: captureFn});
  const screenShotButton = createCaptureButton({label: 'Capture screenshot', onClick: takeScreenshot(canvas, screenshotHolder)});
  const messageBox = createMessageBox();

  targetElement.appendChild(joinElements(captureButton, messageBox, canvas, screenShotButton));
}
