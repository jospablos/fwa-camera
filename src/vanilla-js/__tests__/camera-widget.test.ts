import {
  getByText,
  waitFor,
  getByAltText,
} from '@testing-library/dom'
import {startCamera} from '../capture-button';

function mountCameraWidget() {
  const div = document.createElement('div')
  startCamera(div);
  document.body.appendChild(div);
  return document.body;
}

describe('Camera', () => {
  it('should let a user take a screenshot', async () => {
    const container = mountCameraWidget();
  
    // Activate camera
    getByText(container, 'Activate camera').click();
    await waitFor(() =>
      expect(getByText(container, 'Activate camera')).not.toBeVisible()
    );
    // TODO: test video gets assigned what we expect from the stream.
    
    const screenshot = getByAltText(container, 'screenshot');
    expect(screenshot).not.toBeVisible();
    
    getByText(container, 'take screenshot').click();
    expect(screenshot).toBeVisible();
    expect(screenshot).toMatchSnapshot();
    
    getByText(container, 'clear screenshot').click();
    expect(screenshot).not.toBeVisible();

    // takes a different screenshot every time.
    getByText(container, 'take screenshot').click();
    expect(screenshot).toMatchSnapshot();
  })
})