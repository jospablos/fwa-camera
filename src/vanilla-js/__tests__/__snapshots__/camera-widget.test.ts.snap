// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera Widget should let a user take a screenshot 1`] = `
<img
  alt="screenshot"
  class="centered"
  src="data:image/jpeg;base64,00"
/>
`;

exports[`Camera Widget should let a user take a screenshot 2`] = `
<img
  alt="screenshot"
  class="centered"
  src="data:image/jpeg;base64,01"
/>
`;
